### PR TITLE
Rover add proximity

### PR DESCRIPTION
**Pull request by khancyr**
based on https://github.com/ArduPilot/ardupilot/pull/6775
need InertialNav removal

### CODE REVIEW

**Comment by amilcarlucas** [comment]
Any progree on this one ? Why is it failing Travis ?

**Comment by khancyr** [comment]
@amilcarlucas it need the removal of inertial nav

**Comment by amilcarlucas** [comment]
I miss the link here to the InertialNav PR

File: APMrover2/system.cpp [hunk 1]
```diff
@@ -386,3 +396,70 @@ bool Rover::disarm_motors(void)
...
+// ekf_position_ok - returns true if the ekf claims it's horizontal absolute position estimate is ok and home position is set
+bool Rover::ekf_position_ok()
+{
+    if (!ahrs.have_inertial_nav()) {
+        // do not allow navigation with dcm position
+        return false;
+    }
+
+    // with EKF use filter status and ekf check
+    nav_filter_status filt_status = inertial_nav.get_filter_status();
```
Comment (amilcarlucas) [inline APMrover2/system.cpp]: This adds a dependenvy on inertialNav

File: APMrover2/system.cpp [hunk 2]
```diff
@@ -386,3 +396,70 @@ bool Rover::disarm_motors(void)
...
+    }
+#endif
+#if VISUAL_ODOMETRY_ENABLED == ENABLED
+    if (!g2.visual_odom.enabled()) {
+        return false;
+    }
+#endif
+
+    // get filter status from EKF
+    nav_filter_status filt_status = inertial_nav.get_filter_status();
```
Comment (amilcarlucas) [inline APMrover2/system.cpp]: This adds a dependency on inertial Nav

**Comment by amilcarlucas** [comment]
`
@@ -412,7 +412,10 @@ bool Rover::ekf_position_ok()
     }
 
     // with EKF use filter status and ekf check
-    nav_filter_status filt_status = inertial_nav.get_filter_status();
+    nav_filter_status filt_status;
+    if (!ahrs.get_filter_status(filt_status)) {
+        return false;
+    }
 
     // if disarmed we accept a predicted horizontal position
     if (!arming.is_armed()) {
@@ -447,7 +450,10 @@ bool Rover::optflow_position_ok()
 #endif
 
     // get filter status from EKF
-    nav_filter_status filt_status = inertial_nav.get_filter_status();
+    nav_filter_status filt_status;
+    if (!ahrs.get_filter_status(filt_status)) {
+        return false;
+    }
 
     // if disarmed we accept a predicted horizontal relative position
     if (!arming.is_armed()) {
`

**Comment by khancyr** [comment]
rebased.
Need test

**Comment by rmackay9** [comment]
I'm looking at this now.  One general thing is I would prefer if we don't make copies of the ahrs state in the vehicle code.  We have an ahrs object so we should be able to get what we need directly from it.  making local copies makes me worry about race conditions.

**Comment by OXINARF** [comment]
> One general thing is I would prefer if we don't make copies of the ahrs state in the vehicle code. We have an ahrs object so we should be able to get what we need directly from it. making local copies makes me worry about race conditions.

@rmackay9 We should definitely discuss this at the next dev call then. @khancyr and I have been doing the work for the removal of InertialNav based on this AHRS state being in the vehicles. That is what InertialNav did and there is a good reason to keep it in vehicles: the AHRS doesn't have an API that clearly states it will return the last good values, most methods just return if the values are valid or not. AFAIK no vehicle enters into a failsafe because in one loop the AHRS couldn't provide one value, but that would need to happen if we don't have the last good value stored.

I think that in the future, we should have a stable and standard AHRS API that always gives the last known good values, even if it has to save them. Another possibility lightly discussed in #6063 is that AHRS has methods giving an estimate of the error (and getter methods can then always return something instead of the validity flag).

I'm not sure what the race concern is about though, this all run in the same thread.

**Comment by khancyr** [comment]
Merged by Randy

**Pull request closed without merging**